nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java

```diff
@@ -0,0 +1,383 @@
+    @OnScheduled
+    public void onScheduled() {
+        initialized.set(false);
+    }
+
+    private synchronized void initializeSMTPServer(final ProcessContext context) throws Exception {
+        if (initialized.get()) {
+            return;
+        }
+
+        messages = new LinkedBlockingQueue<>(1024);
+        String clientAuth = null;
+
+        new SMTPMessageHandlerFactory(messages);
+
+        // If an SSLContextService was provided then create an SSLContext to pass down to the server
+        SSLContext sslContext = null;
+        final SSLContextService sslContextService = context.getProperty(SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
+        if (sslContextService != null) {
+            clientAuth = context.getProperty(CLIENT_AUTH).getValue();
+            sslContext = sslContextService.createSSLContext(SSLContextService.ClientAuth.valueOf(clientAuth));
+        }
+
+        final SSLContext finalSslContext = sslContext;
+        final String finalClientAuth = clientAuth;
+
+        SMTPMessageHandlerFactory smtpMessageHandlerFactory = new SMTPMessageHandlerFactory(messages);
+        final SMTPServer server = new SMTPServer(smtpMessageHandlerFactory) {
+
+            @Override
+            public SSLSocket createSSLSocket(Socket socket) throws IOException {
+                InetSocketAddress remoteAddress = (InetSocketAddress) socket.getRemoteSocketAddress();
+
+                SSLSocketFactory socketFactory = finalSslContext.getSocketFactory();
+
+                SSLSocket s = (SSLSocket) (socketFactory.createSocket(socket, remoteAddress.getHostName(), socket.getPort(), true));
+
+                s.setUseClientMode(false);
+
+
+                // For some reason the createSSLContext above is not enough to enforce
+                // client side auth
+                // If client auth is required...
+                if (finalClientAuth == "REQUIRED") {
+                    s.setNeedClientAuth(true);
+                }
+
+
+                return s;
+            }
+        };
+
+
+        // Set some parameters to our server
+        server.setSoftwareName("");
+
+
+        // Set the Server options based on properties
+        server.setPort(context.getProperty(SMTP_PORT).asInteger());
+        server.setHostName(context.getProperty(SMTP_HOSTNAME).getValue());
+        server.setMaxMessageSize(context.getProperty(SMTP_MAXIMUM_MSG_SIZE).asDataSize(DataUnit.B).intValue());
+        server.setMaxConnections(context.getProperty(SMTP_MAXIMUM_CONNECTIONS).asInteger());
+        server.setConnectionTimeout(context.getProperty(SMTP_TIMEOUT).asInteger());
+
+        // Check if TLS should be enabled
+        if (sslContextService != null) {
+            server.setEnableTLS(true);
+        } else {
+            server.setHideTLS(true);
+        }
+
+        // Set TLS to required in case CLIENT_AUTH = required
+        if (context.getProperty(CLIENT_AUTH).getValue() == "REQUIRED") {
```

Comment:
For String comparison, equals() method is recommended instead of '==' operator, since '==' fails with different instance even if it has same value.

```
"foo" == "foo" // true
"foo" == new String("foo") // false
```

An alternative statement would be:
```
SSLContextService.ClientAuth.REQUIRED.toString()
  .equals(context.getProperty(CLIENT_AUTH).getValue())
```